src/R2N.jl

```diff
@@ -0,0 +1,387 @@
+- `σmin = eps(T)`: step parameter for R2N algorithm.
+- `max_eval::Int = -1`: maximum number of evaluation of the objective function.
+- `max_time::Float64 = 30.0`: maximum time limit in seconds.
+- `max_iter::Int = typemax(Int)`: maximum number of iterations.
+- `β = T(0) ∈ [0,1]` is the constant in the momentum term. If `β == 0`, R2N does not use momentum.
+- `verbose::Int = 0`: if > 0, display iteration details every `verbose` iteration.
+- `subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = ShiftedLBFGSSolver`: the subsolver to solve the shifted system. Default is `JSOSolvers.ShiftedLBFGSSolver` which is the exact solver.
+- `subsolver_verbose::Int = 0`: if > 0, display iteration information every `subsolver_verbose` iteration of the subsolver if CG is selected.
+
+# Output
+The value returned is a `GenericExecutionStats`, see `SolverCore.jl`.
+
+# Callback
+The callback is called at each iteration.
+The expected signature of the callback is `callback(nlp, solver, stats)`, and its output is ignored.
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.x`: current iterate;
+- `solver.gx`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.dual_feas`: norm of current gradient;
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
+
+# Examples
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+stats = R2N(nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+
+```jldoctest
+using JSOSolvers, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3))
+solver = R2NSolver(nlp);
+stats = solve!(solver, nlp)
+
+# output
+
+"Execution stats: first-order stationary"
+```
+"""
+mutable struct R2NSolver{
+  T,
+  V,
+  Op <: AbstractLinearOperator{T},
+  Op2 <: AbstractLinearOperator{T},
+  Sub <: Union{KrylovSolver{T, T, V}, ShiftedLBFGSSolver},
+} <: AbstractOptimizationSolver
+  x::V
+  cx::V
+  gx::V
+  gn::V
+  σ::T
+  μ::T
+  H::Op
+  shifted_H::Op2
+  Hs::V
+  s::V
+  obj_vec::V # used for non-monotone behaviour
+  subsolver_type::Sub
+  cgtol::T
+end
+
+function R2NSolver(
+  nlp::AbstractNLPModel{T, V};
+  non_mono_size = 1,
+  subsolver_type::Union{Type{<:KrylovSolver}, Type{ShiftedLBFGSSolver}} = MinresSolver,
+) where {T, V}
+  nvar = nlp.meta.nvar
+  x = V(undef, nvar) # vs similar(nlp.meta.x0)
```

Comment:
```suggestion
  x = V(undef, nvar)
```

Reply:
It doesn't look like you applied the suggested change.